### PR TITLE
Auto detect IMAP special folders when configuring account

### PR DESCRIPTION
**Pull request by Yannick243**
## PR
After [https://github.com/cypht-org/cypht/pull/625](https://github.com/cypht-org/cypht/pull/625), cypht was unable to auto detect IMAP special folders (Trash/Sent/Drafts) even though they exist on the server.
Based on "RFC 6154: IMAP LIST Extension for Special-Use Mailboxes"

### Related issue:
[https://github.com/cypht-org/cypht/issues/817](https://github.com/cypht-org/cypht/issues/817)

### CODE REVIEW

**Comment by marclaporte** [comment]
> After #625, cypht was unable

I don't remember ever seeing this work before.  #625 arrived in Cypht 2.0.0

Can you confirm if this works OK in Cypht 1.4.x?

Thanks!

**Comment by Yannick243** [comment]
The feature was introduced here: [https://github.com/cypht-org/cypht/pull/586](https://github.com/cypht-org/cypht/pull/586)
After testing on version 1.4.x, it appears that it was working partially (on some folders).

With this adjustment to 2.0.0, it is working with all folders when using the configuration form below.
CleanShot 2024-05-14 at 20 52 11@2x
As we can see the folders are automatically detected:
CleanShot 2024-05-14 at 20 53 08@2x

For now, the remaining part that is not working is when using the advanced configuration form below:
[https://imgur.com/a/HW4yFZZ](https://imgur.com/a/HW4yFZZ)

**Comment by marclaporte** [comment]
Ok, but is it working because we are matching folder names or via "RFC 6154: IMAP LIST Extension for Special-Use Mailboxes"?

**Comment by marclaporte** [comment]
@josaphatim any wisdom?

**Comment by josaphatim** [comment]
> Ok, but is it working because we are matching folder names or via "RFC 6154: IMAP LIST Extension for Special-Use Mailboxes"?

> @josaphatim any wisdom?

RFC 6154 says that: This extension adds new optional mailbox attributes that a server may include in IMAP LIST command responses to identify special-use mailboxes to the client. This being said, special-use mailboxes will not always be present for all providers and some will only provide a few special-use mailboxes.

**Comment by marclaporte** [comment]
Ok. So @Yannick243: please make sure to leverage RFC 6154 and not just folder names.

**Comment by marclaporte** [comment]
Sounds good to me. @josaphatim?

**Comment by josaphatim** [comment]
@Yannick243 why this still in draft ?

**Comment by Yannick243** [comment]
This is not working when using the new advanced configuration form [https://github.com/cypht-org/cypht/pull/855](https://github.com/cypht-org/cypht/pull/855). 
That's what I am fixing.

**Comment by Yannick243** [comment]
The PR is ready for a review.

File: modules/core/site.js [hunk 1]
```diff
@@ -2188,34 +2188,49 @@ function submitSmtpImapServer() {
...
-            $("#srv_setup_stepper_enable_sieve").prop('checked', false);
-            $("#srv_setup_stepper_only_jmap").prop('checked', false);
-            $('#step_config-imap_bloc').show();
-            $('#step_config-smtp_bloc').show();
-            $('#srv_setup_stepper_profile_bloc').show();
-
-            Hm_Utils.set_unsaved_changes(1);
-            Hm_Folders.reload_folders(true);
-            location.reload();
+            if (res.nux_server_id) {
```
Comment (josaphatim) [inline modules/core/site.js:2191]: @Yannick243 just have some requests

Can you rename nux_server_id and nux_service_name as we are not in nux module here ?

Also what if imap_folders module is not enabled ? Can fix also for the nux module ?

Last thing we had a confirmation dialog before setting folders https://github.com/cypht-org/cypht/commit/ac3b9c283b8b8c5094f7362adacd8e0368c1c9d6#diff-3e95247a6dbbf437a9b5e7341db84da72c039b7b5ad532928b4cd8678dd5ff17R45. Shall we keep it for both or remove it for nux module too ? @marclaporte can you advise on this ? For my hand I'd prefer to remove it for nux module.

Comment (marclaporte) [reply modules/core/site.js:2191]: I also prefer we remove this confirmation dialog. People can easily change the folders after if they wish (which is super rare).

Comment (Yannick243) [reply modules/core/site.js]: Hello @josaphatim,
The changes have been applied.

File: modules/core/handler_modules.php [hunk 2]
```diff
@@ -1103,6 +1103,9 @@ public function process() {
 
                 Hm_Msgs::add("Server saved");
                 $this->out('just_saved_credentials', true);
+                $this->out('imap_server_id', $this->imap_server_id);
```
Comment (josaphatim) [inline modules/core/handler_modules.php:1106]: @Yannick243 we can avoid adding imap_folders_enabled in allowed output as we don't really need it in js with something like:

```
if ($this->module_is_supported('imap_folders')) {
    $this->out('imap_server_id', $this->imap_server_id);
    $this->out('imap_service_name', $provider);
}
```

File: modules/core/site.js [hunk 3]
```diff
@@ -2188,34 +2188,49 @@ function submitSmtpImapServer() {
...
-            $("#srv_setup_stepper_enable_sieve").prop('checked', false);
-            $("#srv_setup_stepper_only_jmap").prop('checked', false);
-            $('#step_config-imap_bloc').show();
-            $('#step_config-smtp_bloc').show();
-            $('#srv_setup_stepper_profile_bloc').show();
-
-            Hm_Utils.set_unsaved_changes(1);
-            Hm_Folders.reload_folders(true);
-            location.reload();
+            if (res.imap_server_id && res.imap_folders_enabled) {
```
Comment (josaphatim) [inline modules/core/site.js:2191]: Also change this to `if (res.imap_server_id)`

File: modules/imap_folders/modules.php [hunk 4]
```diff
@@ -129,6 +129,12 @@ public function process() {
  */
 class Hm_Handler_process_accept_special_folders extends Hm_Handler_Module {
     public function process() {
+        
+        if (!$this->module_is_supported('imap_folders')) {
+            Hm_Msgs('ERRIMAP folders module is not enabled');
```
Comment (josaphatim) [inline modules/imap_folders/modules.php:134]: I think we can just return as users won't have a clue on this or add it or debug instead.

File: modules/nux/site.js [hunk 5]
```diff
@@ -47,16 +47,18 @@ var nux_add_account = function() {
 
 var display_final_nux_step = function(res) {
     if (res.nux_account_added) {
-        if (confirm('Do you accept special folders?')) {
+        if (res.nux_server_id && res.nux_imap_folders_enabled) {
```
Comment (josaphatim) [inline modules/nux/site.js:50]: Same here

Comment (Yannick243) [reply modules/nux/site.js]: Done

**Pull request merged**